linera-chain/src/certificate.rs

```diff
@@ -0,0 +1,307 @@
+    /// Returns reference to the value contained in this certificate.
+    pub fn inner(&self) -> &T {
+        self.value.inner()
+    }
+
+    /// Consumes this certificate, returning the value it contains.
+    pub fn into_inner(self) -> T {
+        self.value.value
+    }
+}
+
+impl<T: Clone> Clone for CertificateT<T> {
+    fn clone(&self) -> Self {
+        Self {
+            value: self.value.clone(),
+            round: self.round,
+            signatures: self.signatures.clone(),
+        }
+    }
+}
+
+impl<T: Debug> Debug for CertificateT<T> {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("CertificateT")
+            .field("value", &self.value)
+            .field("round", &self.round)
+            .field("signatures", &self.signatures)
+            .finish()
+    }
+}
+
+// NOTE: For backwards compatiblity reasons we serialize the new `Certificate` type as the old
+// one. Otherwise we would be breaking the RPC API schemas. We can't implement generic serialization
+// for `Certificate<T>` since only specific `T`s have corresponding `CertificateValue` variants.
+impl Serialize for ValidatedBlockCertificate {
+    fn serialize<S: Serializer>(&self, serializer: S) -> Result<S::Ok, S::Error> {
+        let cert = Certificate::from(self.clone());
+        cert.serialize(serializer)
+    }
+}
+
+impl<'de> Deserialize<'de> for ValidatedBlockCertificate {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        Certificate::deserialize(deserializer).map(ValidatedBlockCertificate::from)
+    }
+}
+
+#[cfg(with_testing)]
+impl<T: Eq + PartialEq> Eq for CertificateT<T> {}
+#[cfg(with_testing)]
+impl<T: Eq + PartialEq> PartialEq for CertificateT<T> {
+    fn eq(&self, other: &Self) -> bool {
+        self.value.hash == other.value.hash
+            && self.round == other.round
+            && self.signatures == other.signatures
+    }
+}
+
+impl From<ConfirmedBlockCertificate> for Certificate {
+    fn from(cert: ConfirmedBlockCertificate) -> Certificate {
+        let ConfirmedBlockCertificate {
+            value,
+            round,
+            signatures,
+        } = cert;
+        Certificate::new(
+            HashedCertificateValue::new_confirmed(value.into_inner().into_inner()),
+            round,
+            signatures,
+        )
+    }
+}
+
+impl From<ValidatedBlockCertificate> for Certificate {
+    fn from(cert: ValidatedBlockCertificate) -> Certificate {
+        let ValidatedBlockCertificate {
+            value,
+            round,
+            signatures,
+        } = cert;
+        Certificate::new(
+            HashedCertificateValue::new_validated(value.into_inner().into_inner()),
+            round,
+            signatures,
+        )
+    }
+}
+
+// In practice, it should be HashedCertificateValue = Hashed<CertificateValue>
+// but [`HashedCertificateValue`] is used in too many places to change it now.
```

Comment:
We should create issues for the follow-up tasks.

Reply:
Done https://github.com/linera-io/linera-protocol/issues/2842. And added to comment.